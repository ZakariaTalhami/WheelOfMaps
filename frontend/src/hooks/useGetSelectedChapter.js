import { useSelector } from "react-redux";

/**
 * Get the selected chapter instance from redux store
 * @returns {object} chapter
 */
const useGetSelectedChapter = () => {
    const getSelectedChapter = (bookState) => {
        let selectedChapter;

        if (bookState.selectedChapter !== undefined) {
            const chapters = bookState.books[bookState.selectedBook].chapters;
            selectedChapter = chapters.find(
                (chapter) => chapter.number === bookState.selectedChapter
            );
        }

        return selectedChapter;
    };

    return useSelector((state) => getSelectedChapter(state.Books));
};

export default useGetSelectedChapter;
