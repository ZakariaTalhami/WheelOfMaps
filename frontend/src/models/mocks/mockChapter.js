import Chapter from "../chapter";

export const MOCK_BOOK_ID = "<bookID>";
export const MOCK_CHAPTER = {
    _id: "be7983c1-00db-4b1d-aef7-0dfc5946f8e3",
    number: 1,
    title: "title",
    summary: {
        body: "summary",
        link: "link/to/summary",
        author: "author of summary",
    },
    chapterIndex: "10101",
};

export default Chapter.ConstructFromObject(MOCK_BOOK_ID, MOCK_CHAPTER);
