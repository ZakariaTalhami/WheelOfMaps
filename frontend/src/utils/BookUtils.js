/*
    Uility module for common functionality for Books and chapters.
*/

export const getBookChapterTitles = (book) => {
    const chapters = book.chapters || [];
    return chapters.map((chapter) => {
        const prefix =
            chapter.number === 0 ? "Prologue" : `Chapter ${chapter.number}`;

        return `${prefix}: ${chapter.title}`;
    });
};
