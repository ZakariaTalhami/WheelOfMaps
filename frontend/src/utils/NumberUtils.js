/**
 * Pads a number depending on the number of digits required
 * Returns a string with the number padded with zeros
 *
 * @param {Number} num - the desired number
 * @param {Number} places - number of digits
 * @returns {String} formated string with zero padding
 */
export function zeroPad(num, places) {
    var zero = places - num.toString().length + 1;
    return Array(+(zero > 0 && zero)).join("0") + num;
}
