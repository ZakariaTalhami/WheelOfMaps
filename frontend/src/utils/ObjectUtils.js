import _ from "lodash";

const evaluateObjectPath = (obj, path = [], defaultValue) => {
    if (_.isString(path)) path = _.split(path, ".");
    if (!_.isObject(obj) || _.isEmpty(path)) return obj || defaultValue;

    return evaluateObjectPath(obj[_.head(path)], _.drop(path), defaultValue);
};

const ObjectUtils = {
    evaluateObjectPath,
};

export default ObjectUtils;
