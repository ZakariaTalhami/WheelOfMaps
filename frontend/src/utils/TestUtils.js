import React from "react";
import { render } from "@testing-library/react";
import { ChakraProvider } from "@chakra-ui/react";
//Global Theme
import extendedTheme from "../assets/styles/extendedTheme";
// TODO: remove custom font icons, replace with react-icons
import "../assets/fonts/style.css";
// Leafletjs Styles
import "leaflet/dist/leaflet.css";
import "../assets/_leafletjs.override.scss";

const AllTheProviders = ({ children }) => {
    Object.defineProperty(window, "matchMedia", {
        writable: true,
        value: jest.fn().mockImplementation((query) => ({
            matches: false,
            media: query,
            onchange: null,
            addListener: jest.fn(), // deprecated
            removeListener: jest.fn(), // deprecated
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        })),
    });

    return <ChakraProvider theme={extendedTheme}>{children}</ChakraProvider>;
};

const customRender = (ui, options) =>
    render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
