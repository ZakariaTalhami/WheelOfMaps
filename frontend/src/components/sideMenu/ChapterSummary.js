// Core
import React from "react";
//Components
import { Center, Heading, Text, VStack } from "@chakra-ui/react";
// Hooks
import useGetSelectedChapter from "../../hooks/useGetSelectedChapter";
// Utils
import { getChapterTitle } from "../../utils/BookUtils";

/**
 * not chpater selected message component
 */
const NoChapterSummarySelected = () => (
    <Center>
        <Heading size="lg">No Chapter selected</Heading>
    </Center>
);

const ChapterHeader = (props) => <Heading size="md" {...props} />;

const ChpaterSummaryWrapper = (props) => (
    <VStack p="1rem" overflow="auto" textAlign="justify" {...props} />
);

/**
 * Split the summary by new line and wrap in paragraphs
 * @param {String} summary
 * @returns {Array} paragraphs of the summary
 */
const getSummaryParagraphs = (summary) => {
    const paragraphs = summary.split("\r\n\r\n");
    return paragraphs.map((para, index) => (
        <Text style={{ textIndent: "1.5rem" }} key={index}>
            {para}
        </Text>
    ));
};

const ChapterSummary = () => {
    const chapter = useGetSelectedChapter();

    return (
        <>
            {!chapter ? (
                <NoChapterSummarySelected />
            ) : (
                <VStack h="100%">
                    <ChapterHeader>{getChapterTitle(chapter)}</ChapterHeader>
                    <ChpaterSummaryWrapper>
                        {getSummaryParagraphs(chapter.summary)}
                    </ChpaterSummaryWrapper>
                </VStack>
            )}
        </>
    );
};

export default ChapterSummary;
