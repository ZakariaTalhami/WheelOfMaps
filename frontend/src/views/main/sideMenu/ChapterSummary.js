// Core
import React from "react";
//Components
import { Center, Heading, VStack } from "@chakra-ui/react";
import { DrawerHeader, DrawerContentWrapper } from "./drawer/DrawerContent";
import CreditLink from "../../../components/creditLink/CreditLink";
// Hooks
import useGetSelectedChapter from "../../../hooks/useGetSelectedChapter";
// Utils
import { getSummaryParagraphs } from "../../../utils/TextUtils";
import { getChapterTitle } from "../../../utils/BookUtils";

/**
 * not chpater selected message component
 */
const NoChapterSummarySelected = () => (
    <Center>
        <Heading size="lg">No Chapter selected</Heading>
    </Center>
);

const ChapterSummary = () => {
    const chapter = useGetSelectedChapter();
    const summary = chapter?.summary;
    return (
        <>
            {!chapter ? (
                <NoChapterSummarySelected />
            ) : (
                <VStack h="100%">
                    <DrawerHeader>{getChapterTitle(chapter)}</DrawerHeader>
                    <DrawerContentWrapper>
                        {getSummaryParagraphs(summary.body)}
                        <CreditLink
                            author={summary.author}
                            link={summary.link}
                        />
                    </DrawerContentWrapper>
                </VStack>
            )}
        </>
    );
};

export default ChapterSummary;
