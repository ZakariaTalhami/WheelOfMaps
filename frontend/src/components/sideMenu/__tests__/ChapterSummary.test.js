// Libs
import { render, screen } from "../../../utils/TestUtils";
// Tested component
import ChapterSummary from "../ChapterSummary";
// Mocks
import useGetSelectedChapter from "../../../hooks/useGetSelectedChapter";
import mockChapter from "../../../models/mocks/mockChapter";
// Utils
import { getChapterTitle } from "../../../utils/BookUtils";

jest.mock("../../../hooks/useGetSelectedChapter");

describe("Selected Chapter is undefined", () => {
    beforeEach(() => {
        useGetSelectedChapter.mockReturnValue(undefined);
    });

    test("Show no selection message", () => {
        render(<ChapterSummary />);

        expect(screen.getByText("No Chapter selected")).toBeDefined();
    });

    test("Match snapshot", () => {
        const { container } = render(<ChapterSummary />);

        expect(container).toMatchSnapshot();
    });
});

describe("Selected Chapter is defined", () => {
    beforeEach(() => {
        useGetSelectedChapter.mockReturnValue(mockChapter);
    });

    test("show chapter summary", () => {
        render(<ChapterSummary />);

        expect(screen.getByText(getChapterTitle(mockChapter))).toBeDefined();
        expect(screen.getByText(mockChapter.summary)).toBeDefined();
    });

    test("Match snapshot when selected chapter is undefined", () => {
        const { container } = render(<ChapterSummary />);

        expect(container).toMatchSnapshot();
    });
});
