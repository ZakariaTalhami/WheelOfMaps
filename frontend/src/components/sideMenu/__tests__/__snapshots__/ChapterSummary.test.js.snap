// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selected Chapter is defined Match snapshot when selected chapter is undefined 1`] = `
<div>
  <div
    class="chakra-stack css-10e2phi"
  >
    <h2
      class="chakra-heading css-1chwp2w"
    >
      Chapter 1: title
    </h2>
    <div
      class="chakra-stack css-tlrjpm"
    >
      <p
        class="chakra-text css-1u8qly9"
        style="text-indent: 1.5rem;"
      >
        summary
      </p>
    </div>
  </div>
</div>
`;

exports[`Selected Chapter is undefined Match snapshot 1`] = `
<div>
  <div
    class="css-d85ps1"
  >
    <h2
      class="chakra-heading css-rq7u43"
    >
      No Chapter selected
    </h2>
  </div>
</div>
`;
